hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestSnapshotBackgroundServices.java

```diff
@@ -0,0 +1,692 @@
+
+    // create snapshot b
+    SnapshotInfo snapshotInfoB = createOzoneSnapshot(newLeaderOM,
+        SNAPSHOT_NAME_PREFIX + RandomStringUtils.randomNumeric(5));
+    Assertions.assertNotNull(snapshotInfoB);
+
+    // delete key a
+    ozoneBucket.deleteKey(keyNameA);
+
+    LambdaTestUtils.await(10000, 1000,
+        () -> !isKeyInTable(keyA, omKeyInfoTable));
+
+    // create snapshot c
+    SnapshotInfo snapshotInfoC = createOzoneSnapshot(newLeaderOM,
+        SNAPSHOT_NAME_PREFIX + RandomStringUtils.randomNumeric(5));
+
+    // get snapshot c
+    OmSnapshot snapC;
+    try (ReferenceCounted<IOmMetadataReader, SnapshotCache> rcC = newLeaderOM
+        .getOmSnapshotManager()
+        .checkForSnapshot(volumeName, bucketName,
+            getSnapshotPrefix(snapshotInfoC.getName()), true)) {
+      Assertions.assertNotNull(rcC);
+      snapC = (OmSnapshot) rcC.get();
+    }
+
+    // assert that key a is in snapshot c's deleted table
+    LambdaTestUtils.await(10000, 1000,
+        () -> isKeyInTable(keyA, snapC.getMetadataManager().getDeletedTable()));
+
+    // create snapshot d
+    SnapshotInfo snapshotInfoD = createOzoneSnapshot(newLeaderOM,
+        SNAPSHOT_NAME_PREFIX + RandomStringUtils.randomNumeric(5));
+
+    // delete snapshot c
+    client.getObjectStore()
+        .deleteSnapshot(volumeName, bucketName, snapshotInfoC.getName());
+
+    LambdaTestUtils.await(30000, 1000, () ->
+        !isKeyInTable(snapshotInfoC.getTableKey(),
+            newLeaderOM.getMetadataManager().getSnapshotInfoTable()));
+
+    // get snapshot d
+    OmSnapshot snapD;
+    try (ReferenceCounted<IOmMetadataReader, SnapshotCache> rcD = newLeaderOM
+        .getOmSnapshotManager()
+        .checkForSnapshot(volumeName, bucketName,
+            getSnapshotPrefix(snapshotInfoD.getName()), true)) {
+      Assertions.assertNotNull(rcD);
+      snapD = (OmSnapshot) rcD.get();
+    }
+
+    // wait until key a appears in deleted table of snapshot d
+    LambdaTestUtils.await(10000, 1000,
+        () -> isKeyInTable(keyA, snapD.getMetadataManager().getDeletedTable()));
+
+    // Confirm entry for deleted snapshot removed from info table
+    client.getObjectStore()
+        .deleteSnapshot(volumeName, bucketName, newSnapshot.getName());
+    LambdaTestUtils.await(10000, 1000,
+        () -> isKeyInTable(newSnapshot.getTableKey(),
+            newLeaderOM.getMetadataManager().getSnapshotInfoTable()));
+
+    confirmSnapDiffForTwoSnapshotsDifferingBySingleKey(
+        newLeaderOM);
+  }
+
+  private static <V> boolean isKeyInTable(String key, Table<String, V> table) {
+    try (TableIterator<String, ? extends Table.KeyValue<String, V>> iterator
+             = table.iterator()) {
+      while (iterator.hasNext()) {
+        Table.KeyValue<String, V> next = iterator.next();
+        if (next.getKey().contains(key)) {
+          return true;
+        }
+      }
+      return false;
+    } catch (IOException e) {
+      return false;
```

Comment:
nit: Should we just throw `IOException` and let the caller decide how to deal with it? I'm not sure if exception should be treated as `false`. Not a big concern tho.